### DISCUSSION TITLE
What happens if i insert data with the same id several times?

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [milvus-io/milvus#18201] [Q&A and General discussion] [answered]
What happens if i insert data with the same id several times?
Will it be overriten or it will still be in collection but you only last entry with the same will be accessible>

**Comment by yhmo** [comment, marked as answer]
Milvus doesn't check duplicate primary keys, which means you can insert the same primary keys in the same collection.
If you inserted 100 entities with a same primary key "999"
- you call query(expr="id==999"), milvus will return the first entities he found to you.
- you call search(), the results may contain lot of items with same primary key "999".

**Reply by Goooooooooooooooogle** [reply]
Hi, How can I call search() by using 'id==999'?

**Reply by yhmo** [reply]
> Hi, How can I call search() by using 'id==999'?

Documentation: https://milvus.io/docs/v2.1.x/hybridsearch.md

Example:
Assume you have a collection with two fields: 
- one is for ID, field name is "id"
- the other is a vector field, field name is "vec"

Use an expression within a scope. Here we set "expr" : "id < 1000", the search engine will firstly filter out items whose "id" is less than 1000, then do ANN search in these items.
```
search_param = {
  "data": target_vectors,
  "anns_field": "vec",
  "param": {"metric_type": "L2", "params": {"nprobe": 10}},
  "limit": 10,
  "expr": "id <= 1000",
}
res = collection.search(**search_param)
```

If you set "expr":"id=999" to call search(), the search scope will only contain one item.

**Comment by g-laz77** [comment]
@yhmo why is the entry not overwritten if the primary key is the same?
Is there some design reason why this cannot be done?

**Reply by xiaofan-luan** [reply]
this is the internal implementation since old id and new id is not in the same segment.

We introduce another operation as upsert which will delete the original entity and insert a new one

**Reply by ndtreviv** [reply]
Am I correct in assuming that an upsert updates something if it's there or creates it if not?

If so, what's stopping me from upserting everything all the time? I'm assuming there's a performance penalty...?

**Comment by yhmo** [comment]
> Am I correct in assuming that an upsert updates something if it's there or creates it if not?
Yes, correct

> 
> If so, what's stopping me from upserting everything all the time? I'm assuming there's a performance penalty...?
call upsert() if you know the id is possibly duplicated in the db, otherwise, call insert(). upsert() check the id existence first, while insert() doesn't do that.
